drivers/serial/uart_native_posix.c

```diff
@@ -234,33 +398,515 @@ static int np_uart_0_init(const struct device *dev)
+		if (DEBUG(dev)) {
+			LOG("UART_%d RX not ready\n", INDEX(dev));
+		}
+		return 0;
+	}
+	if (DEBUG(dev)) {
+		LOG("UART_%d RX ok\n", INDEX(dev));
+	}
+	return 1;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static int np_uart_irq_rx_ready(const struct device *dev)
+{
+	struct native_uart_status *d = (struct native_uart_status *)dev->data;
+
+	return np_uart_rx_ready(dev) && d->rxen;
+}
+
+static int np_uart_irq_tx_ready(const struct device *dev)
+{
+	struct native_uart_status *d = (struct native_uart_status *)dev->data;
+	struct pollfd pfd = {
+		.fd = d->out_fd,
+		.events = POLLOUT,
+	};
+
+	poll(&pfd, 1, 0);
+	if (!(pfd.revents & POLLOUT)) {
+		/* Not ready */
+		if (DEBUG(dev)) {
+			LOG("UART_%d TX not ready\n", INDEX(dev));
+		}
+		return 0;
+	}
+	if (DEBUG(dev)) {
+		LOG("UART_%d TX ok\n", INDEX(dev));
+	}
+	return d->txen;
+}
+
+static void *np_uart_irq_worker(void *context)
+{
+	const struct device *dev = (const struct device *)context;
+	struct native_uart_status *d = (struct native_uart_status *)dev->data;
+
+	/* Setup wake source */
+	int pipefd[2];
+
+	if (pipe(pipefd) != 0) {
+		EXIT("Failed to instantiate worker signal\n");
+		return NULL;
+	}
+	d->event_fd = pipefd[1];
+
+	/* Keep going forever */
+	for (;;) {
+		/* Fake an interrupt request mechanism using poll() */
+		bool bIrq = false;
+
+		/* Decide what we're waiting for */
+		struct pollfd pfd[3] = {
+			[0] = {
+				.fd = d->in_fd,
+				.events = d->rxen ? POLLIN : 0,
+			},
+			[1] = {
+				.fd = d->out_fd,
+				.events = d->txen ? POLLOUT : 0,
+			},
+			[2] = {
+				.fd = pipefd[0],
+				.events = POLLIN,
+			}
+		};
+
+		/* Wait forever on one of these things */
+		if (DEBUG(dev)) {
+			LOG("UART_%d WFI\n", INDEX(dev));
+		}
+		(void)poll(pfd, 3, -1);
+		if (DEBUG(dev)) {
+			LOG("UART_%d IRQ ", INDEX(dev));
+		}
+
+		/* Handle events */
+		if (pfd[0].revents != 0) {
+			if (pfd[0].revents == POLLIN) {
+				/* Callback should service */
+				bIrq = true;
+				if (DEBUG(dev)) {
+					LOG("RX ");
+				}
+			} else {
+				/* We can't handle POLLERR, POLLHUP, POLLNVAL */
+				EXIT("%s in_fd revent 0x%x\n", __func__, pfd[0].revents);
+			}
+		}
+		if (pfd[1].revents != 0) {
+			if (pfd[1].revents == POLLOUT) {
+				/* Callback should service */
+				bIrq = true;
+				if (DEBUG(dev)) {
+					LOG("TX ");
+				}
+			} else {
+				/* We can't handle POLLERR, POLLHUP, POLLNVAL */
+				EXIT("%s out_fd revent 0x%x\n", __func__, pfd[1].revents);
+			}
+		}
+		if (pfd[2].revents != 0) {
+			if (pfd[2].revents == POLLIN) {
+				/* This just wakes us up to re-read d->rxen, d->txen etc */
+				int dummy;
+
+				dummy = read(pipefd[0], &dummy, 1);
+				if (DEBUG(dev)) {
+					LOG("WAKE ");
+				}
+			} else {
+				/* We can't handle POLLERR, POLLHUP, POLLNVAL */
+				EXIT("%s event_fd revent 0x%x\n", __func__, pfd[2].revents);
+			}
+		}
+		if (DEBUG(dev)) {
+			LOG("\n");
+		}
+
+		/* Wake simulated CPU? */
+		if (bIrq) {
+			d->pending = true;
+			hw_irq_ctrl_set_irq(d->irq);
```

Comment:
The HW models code in native_posix is not meant to be called from asynchronous threads. Meaning, this will break and crash in the most mysterious ways in the most unsuspecting times. Unfortunately this is not the way to go :(

Reply:
Hi @aescolar thanks for the review.

Enablement of interrupt driven UART for native_posix will allow the native posix build to use the newly added LC3 codec on real (not simulated) Bluetooth controller devices for ISO data which is currently blocked by no support in BlueZ.

This brings value to Zephyr, so could you elaborate a bit on what still needs to be fixed for this PR to be useable? I based the code on the Timer driver and we have not observed issues running with these changes, hence are not sure where to start looking - a pointer to an example of how to correctly inject OS events into the native posix HW model would be appreciated.

Reply:
> could you elaborate a bit on what still needs to be fixed for this PR to be useable?

Probably the easiest way would be to create a HW model that runs in the HW models thread (like the timer hw model). That HW model would then be scheduled to run every X ms of simulated time and would poll for ready data (instead having an asynchronous thread waiting for the data)